src/ast/diagnostic.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <cassert>
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "location.hh"
+
+namespace bpftrace {
+namespace ast {
+
+// Diagnostic reflects a single error at a single source location. This is a
+// simple wrapper around a string for that message, and the location class.
+class Diagnostic {
+public:
+  enum class Severity {
+    Warning,
+    Error,
+  };
+  Diagnostic(location loc) : loc_(loc) {};
+  const std::string msg() const
+  {
+    return msg_.str();
+  }
+  const std::string hint() const
+  {
+    return hint_.str();
+  }
+  const location& loc() const
+  {
+    return loc_;
+  }
+
+  // Each diagnostic can potentially have a hint attached, which is how to
+  // effectively resolve this issue.
+  std::stringstream& addHint()
+  {
+    return hint_;
+  }
+
+  template <typename T>
+  Diagnostic& operator<<(const T& t)
+  {
+    msg_ << t;
+    return *this;
+  }
+
+private:
+  std::stringstream msg_;
+  std::stringstream hint_;
+  location loc_;
+};
+
+// Diagnostics contains a collection of diagnostics.
```

Comment:
Not a very helpful comment IMHO :)

Reply:
I find it useless. You could re-word it but as is, I would remove.

Reply:
Candidly, I don't think it was worth the time to iterate on this, but the comment is now gone.